File: 1029-vertical-order-traversal-of-a-binary-tree/1029-vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void dfs(TreeNode root, int hd, int level, TreeMap<Integer, TreeMap<Integer, ArrayList<Integer>>> map) {
        if (root == null) return;

        map.putIfAbsent(hd, new TreeMap<>());
        map.get(hd).putIfAbsent(level, new ArrayList<>());
        map.get(hd).get(level).add(root.val);
        
        dfs(root.left, hd - 1, level + 1, map);  
        dfs(root.right, hd + 1, level + 1, map); 
    }
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        TreeMap<Integer, TreeMap<Integer, ArrayList<Integer>>> map = new TreeMap<>();
        dfs(root, 0, 0, map);
        List<List<Integer>> result = new ArrayList<>();
        
        for (TreeMap<Integer, ArrayList<Integer>> levelMap : map.values()) {
            List<Integer> vertical = new ArrayList<>();
            for (ArrayList<Integer> nodes : levelMap.values()) {
                Collections.sort(nodes); 
                vertical.addAll(nodes);
            }
            result.add(vertical);
        }
        return result;
    }
} 

File: 1029-vertical-order-traversal-of-a-binary-tree/1029-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

    void helper(TreeNode * root, map<int, map<int,vector<int>>> &m, int hd, int level) {
        if(root == nullptr) return;

        m[hd][level].push_back(root -> val);
        helper(root -> left, m, hd - 1, level + 1); 
        helper(root -> right, m, hd + 1, level + 1); 
    }
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        map<int, map<int,vector<int>>> m;
        helper(root, m, 0, 0);

        vector<vector<int>> v1;
        for(auto i : m) {
            vector<int> v2;
            for(auto j : i.second) {
                sort(j.second.begin(),j.second.end());
                for(auto val : j.second) {
                    v2.push_back(val);
                }
            }
            v1.push_back(v2);
        }
        return v1;
    }
};